### PR TITLE
Cluster aware search using Multiviewer

### PR DESCRIPTION
**Pull request by codesniffer**
**Problem Statement**
In the Multiviewer, a user cannot search for sessions by selecting specific Elasticsearch cluster/clusters. In the current implementation, MultiES broadcast a search request to all the ES clusters and aggregates the results. This approach might increase the response time.

**Solution**
1. Allow Multiviewer to display available ES cluster
2. Allow user to select particular cluster/clusters
3. Enhance the capability of Viewer.js, Db.js, MultiES.js, and Vue app to handle them cluster-specific search

## License
I confirm that this contribution is made under an Apache 2.0 license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

**Comment by codesniffer** [comment]
Thanks you for suggestions. They are very helpful. I will make the change accordingly and keep you informed.

**Comment by codesniffer** [comment]
@31453 , I have made the changes and attached the screenshots for your review. Can you please provide your suggestions. I did not push the changes yet.

image

image

image

image

**Comment by codesniffer** [comment]
@awick, it would be great if you have the time to review the images and share your thoughts.

**Comment by codesniffer** [comment]
@31453 , completed a revision with the recommended changes. Please take the time to review and provide your suggestions.

**Comment by 31453** [comment]
Looks great! ✨ This is very cool 😄 I'd like to see a few minor tweaks: 

1. If a single cluster is selected, show that as the title of the dropdown (next to the database icon)
2. If multiple clusters are selected, show how many clusters are being searched
3. If the user hits search, and no clusters are selected, return an error to indicate that no cluster are selected

I see some screenshots of 1 and 2 above, did those not get pushed?

@awick do you think that searching all clusters is a good default? Or is that risky if there are lots of clusters, which would make searching slow? Should there be some sort of configurable default to select one or more clusters to search by default?

**Comment by 31453** [comment]
The Hunt tab is missing from the application and Shortcuts are missing from the Settings page. This is because the server is sending the client the `MOLOCH_MULTIVIEWER` flag as true. While this is technically correct, hunts and shortcuts should work. 

For Hunt, there should probably be a new control on the create hunt form that asks which clusters to hunt.

**Comment by 31453** [comment]
I'm assuming that all users have access to all clusters in this PR. Right now in Arkime, you have to add a user per cluster. Do we want to lock down clusters from certain users? Would we add this to the user settings page?

**Comment by codesniffer** [comment]
> 3\. If the user hits search, and no clusters are selected, return an error to indicate that no cluster are selected

I found that 1 & 2 are missing from the commits. I am working to make those changes.
I will also work on Item 3 and push the changes soon.

**Comment by awick** [comment]
If I remember right for multiviewer it only uses the user config from the first ES listed or usersElasticsearch if set.

I think all clusters is a fine default for now :)

**Comment by codesniffer** [comment]
> Looks great! ✨ This is very cool 😄 I'd like to see a few minor tweaks:
> 
> 1. If a single cluster is selected, show that as the title of the dropdown (next to the database icon)
> 2. If multiple clusters are selected, show how many clusters are being searched
> 3. If the user hits search, and no clusters are selected, return an error to indicate that no cluster are selected
> 
> I see some screenshots of 1 and 2 above, did those not get pushed?
> 
> @awick do you think that searching all clusters is a good default? Or is that risky if there are lots of clusters, which would make searching slow? Should there be some sort of configurable default to select one or more clusters to search by default?

I have completed item 1, 2 and 3, and pushed the changes for your review. Please suggest.

**Comment by codesniffer** [comment]
> The Hunt tab is missing from the application and Shortcuts are missing from the Settings page. This is because the server is sending the client the `MOLOCH_MULTIVIEWER` flag as true. While this is technically correct, hunts and shortcuts should work.
> 
> For Hunt, there should probably be a new control on the create hunt form that asks which clusters to hunt.

@31453 , can you please provide some details for this ask. 
1. Do we want to display the Hunt Tab in Multiviwer?
2. If the Hunt Tab is displayed then  should we display a list of Cluster so that a user can run the Hun job for that cluster?
3. Do we want to include this feature in this release?

**Comment by codesniffer** [comment]
> I'm assuming that all users have access to all clusters in this PR. Right now in Arkime, you have to add a user per cluster. Do we want to lock down clusters from certain users? Would we add this to the user settings page?
Are we consider these as additional feature? According to my understanding, the current implementation of the Multiviwer does not prevent a user to search against a Cluster based on permissions. Should this feature be considered in this PR?

**Comment by 31453** [comment]
@codesniffer Ignore my stuff about hunt and shortcuts! Since this is only applicable to running multiviewer, those items should be hidden and not work 😄 In the future, I would love for these things to work in multiviewer, but that's for another time. 

Also ignore my comment about users, I misunderstood how it currently works. You are correct!

**Comment by codesniffer** [comment]
> @codesniffer Ignore my stuff about hunt and shortcuts! Since this is only applicable to running multiviewer, those items should be hidden and not work 😄 In the future, I would love for these things to work in multiviewer, but that's for another time.
> 
> Also ignore my comment about users, I misunderstood how it currently works. You are correct!

Thank you for the feedback. Appreciate it. 
Currently, a user can select one item from at a time from the Menu. I am wondering whether we should make the ES Cluster Dropdown menu as Multi Select. This way a user can select multiple items. What is your suggestion on that.

**Comment by 31453** [comment]
Yes! I agree. I was just about to suggest that. 😄 

I see the tooltip for selected clusters on the button. Nice work 👍 But I can't seem to trigger an error when unselecting all of the clusters. It is still returning results from all clusters. How are you running this locally for testing? Maybe I'm not running it correctly

**Comment by codesniffer** [comment]
> I see the tooltip for selected clusters on the button. Nice work 👍 But I can't seem to trigger an error when unselecting all of the clusters. It is still returning results from all clusters. How are you running this locally for testing? Maybe I'm not running it correctly

I am selecting Clear All and then clicking the Search Button.

**Comment by codesniffer** [comment]
@31453 , I have provided a fix. Hopefully it will work. The test methods are as follows:

Method 1. Clear the selection and then issue search
Method 2. Set the esquery to none. On the browser type the URL http://localhost:8005/sessions?escluster=none
Method3: Set esquery to an invalid cluster name (e.g., invalid). On the browser type the URL http://localhost:8005/sessions?escluster=invalid

**Comment by 31453** [comment]
That works 👍  Thanks!

**Comment by codesniffer** [comment]
> That works 👍 Thanks!

What are the pending tasks we have at this time? Can you please guide on that?

**Comment by 31453** [comment]
I think it looks great! I'm looking through the code now and will let you know if I have any suggestions.

@awick do you have any thoughts?

**Comment by 31453** [comment]
Are you able to load packets for sessions? I'm not seeing them show up, they are pending until the request time out.

**Comment by codesniffer** [comment]
> Are you able to load packets for sessions? I'm not seeing them show up, they are pending until the request time out.

I am not showing the Packet details. However, the session details are showing fine.

**Comment by codesniffer** [comment]
@31453 ,  can you please try setting the **isLocalViewRegExp** to the name of the nodename https://arkime.com/settings#islocalviewregexp. Put the isLocalViewRegExp setting under the [all] section of the config.ini. This will prevent the Multiviewer to proxy the request.

**Comment by 31453** [comment]
@codesniffer yup, that worked! Thanks! Can you add that to the `testmulties` area of the `tests/config.test.ini` so that it works when running locally with `npm run viewer:testmultiviewer`?

**Comment by codesniffer** [comment]
> @codesniffer yup, that worked! Thanks! Can you add that to the `testmulties` area of the `tests/config.test.ini` so that it works when running locally with `npm run viewer:testmultiviewer`?

Sure, that can be done. However, I would suggest not to add this setting to the tests/config.test.ini. It is expected that the Multiviewer and Viewer run side by side. Can you please confirm that both the Viewer and Multiviewer were running during the test? If the Viewer was running then the Multiviewer would be able to fetch the packet details by proxing the request to the Viewer.

**Comment by codesniffer** [comment]
@31453 , I have added **isLocalViewRegExp=testmulties** to **testmulties** section of the tests/config.test.ini. Please review.

**Comment by 31453** [comment]
👍 looks great! thanks for making that change

File: viewer/vueapp/src/components/spigraph/Spigraph.vue [hunk 1]
```diff
@@ -503,6 +507,22 @@ export default {
       this.loading = true;
       this.error = false;
 
+      if (this.multiviewer) {
+        var availableESCluster = this.$store.state.esCluster.availableCluster.active;
```
Comment (31453) [inline viewer/vueapp/src/components/spigraph/Spigraph.vue:511]: For new code, I'd like to use `let`. This is something we're currently working on as a lot of the old code uses `var`. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let

File: viewer/multies.js [hunk 2]
```diff
@@ -115,15 +118,59 @@ function node2Prefix (node) {
   return '';
 }
 
-var activeESNodes = [];
+function removeClusterFromQueryUrl (url) {
```
Comment (awick) [inline viewer/multies.js:121]: I think this whole function can be replaced with
url = url.replace(/_cluster=[^&]*(&|$)/g, '');

File: tests/config.test.ini [hunk 3]
```diff
@@ -108,6 +108,8 @@ elasticsearch=http://127.0.0.1:8200
 multiES=true
 multiESPort=8200
 multiESNodes=127.0.0.1:9200,prefix:tests;127.0.0.1:9200,prefix:tests2_
+multiESClusters=test;test2
+isLocalViewRegExp=testmulties
```
Comment (awick) [inline tests/config.test.ini:112]: why is this needed?

**Review by awick — CHANGES_REQUESTED** [review]
Code looks good, asked for some minor changes. One general think I don't like is that escluster is used in one direction and in the code, while _cluster is used in the other direction. Seems like should use escluster everywhere maybe?

I still need to try the UI

File: tests/config.test.ini (hunk 3 above)
Comment (31453) [reply tests/config.test.ini:112]: `isLocalViewRegExp` is needed to test multiviewer locally when running `npm run viewer:multiviewer` and `npm run viewer:testmultiviewer`.

Comment (awick) [reply tests/config.test.ini:112]: Hmm ok, something else must be wrong because it shouldn't, I'll take a look.

**Comment by codesniffer** [comment]
Thanks for taking the time to review. I will make the changes soon.

File: viewer/apiSessions.js [hunk 4]
```diff
@@ -897,7 +901,11 @@ module.exports = (Config, Db, decode, internals, molochparser, Pcap, ViewerUtils
     let fixFields = nonArrayFields.filter(function (x) { return fields.indexOf(x) !== -1; });
 
     async.eachLimit(ids, 10, function (id, nextCb) {
-      Db.getSession(id, { _source: fields.join(',') }, function (err, session) {
+      let options = { _source: fields.join(',') };
+      if (req.query.escluster && Config.get('multiES', false)) {
```
[the remark this answers is not in the document]
Comment (codesniffer) [reply viewer/apiSessions.js]: Added a function as per the suggestion.

File: tests/api-files.t [hunk 5]
```diff
@@ -15,6 +15,7 @@ my ($url) = @_;
     my $json = viewerGet($url);
     my $mjson = multiGet($url);
 
+    for (my $i=0; $i < scalar(@{$mjson->{data}}); $i++) { delete $mjson->{data}->[$i]->{escluster}; }
```
[the remark this answers is not in the document]
Comment (codesniffer) [reply tests/api-files.t:22]: Added test cases to validate if _escluster_ field exist for multiGet respone.

File: viewer/vueapp/src/store.js [hunk 6]
```diff
@@ -175,6 +183,15 @@ const store = new Vuex.Store({
         if (i < len - 1) { state.sortsParam += ','; }
       }
     },
+    setAvailableESCluster (state, value) {
+      state.esCluster.availableCluster = value;
+    },
+    setSelectedESCluster (state, value) {
+      state.esCluster.selectedCluster = value;
+    },
+    setMultiEsStatus (state, value) {
```
[the remark this answers is not in the document]
Comment (codesniffer) [reply viewer/vueapp/src/store.js]: Those were unused variables, so deleted them.

File: viewer/vueapp/src/store.js [hunk 7]
```diff
@@ -33,7 +33,15 @@ const store = new Vuex.Store({
...
-    stickySessionsBtn: false
+    stickySessionsBtn: false,
+    esCluster: {
+      availableCluster: {
+        active: [],
+        inactive: []
+      },
+      selectedCluster: []
+    },
+    multiEsEnabled: false
```
[the remark this answers is not in the document]
Comment (codesniffer) [reply viewer/vueapp/src/store.js]: this variable was not used, so deleted it.

File: viewer/apiSessions.js [hunk 8]
```diff
@@ -1032,9 +1040,9 @@ module.exports = (Config, Db, decode, internals, molochparser, Pcap, ViewerUtils
         return nextCb(null);
       }
 
-      let node = (Config.get('multiES', false) && session._node) ? session._node : undefined;
+      let escluster = (Config.get('multiES', false) && session._source.escluster) ? session._source.escluster : undefined;
```
[the remark this answers is not in the document]
Comment (codesniffer) [reply viewer/apiSessions.js]: I have added to both **_source** and to the **top level**. Need to add the "escluser" field to __source so that it can be used to fetch session details.

File: viewer/apiSessions.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (codesniffer) [reply viewer/apiSessions.js]: > Code looks good, asked for some minor changes. One general thing I don't like is that escluster is used in one direction and in the code, while _cluster is used in the other direction. Seems like should use escluster everywhere maybe?
> 
> I still need to try the UI

I used "escluster" everywhere instead. Please where "_cluster" were used were renamed to "escluster"

**Comment by codesniffer** [comment]
@awick , I have completed making the changes that were suggested. Please review these when time permits.

**Comment by awick** [comment]
This looks really cool!

There are 2 major changes I want to talk about:
1) You are really selecting an Arkime Cluster NOT an ES Cluster. I think all text including, all the changes you did for escluster should change back to cluster. Including in the UI. Sorry about that. you can use arkimeCluster if you want instead.
2) I finally understand that multiESClusters is a parallel array to multiESNodes. Wondering why you didn't just add a name:test to multiESNodes and make a new node2Name function? If you don't want to do that, than you need to make sure clusterList.length === nodes.length at startup

Also, I'm still unsure why you need to set _source.escluster as `egrep '(source|fields).escluster' *.js` returns no results other than in multies

minor stuff:
* I think in toplevel package.json, viewer:multiviewer should be renamed viewer:testmulties and instead of -n all it should use -n testmulties
* the reason `isLocalViewRegExp=testmulties` is required is because we are abusing multiviewer, the design was never for multiviewer to have pcap locally, it was always supposed to reach out to real viewers, but I guess its ok for dev until something breaks. As long as the test suite doesn't use it.
* how can i test the inactive part?

**Comment by codesniffer** [comment]
> This looks really cool!
> 
> There are 2 major changes I want to talk about:
> 
> 1. You are really selecting an Arkime Cluster NOT an ES Cluster. I think all text including, all the changes you did for escluster should change back to cluster. Including in the UI. Sorry about that. you can use arkimeCluster if you want instead.

I will make the changes. I will rename **escluster** to **cluster**

> 2. I finally understand that multiESClusters is a parallel array to multiESNodes. Wondering why you didn't just add a name:test to multiESNodes and make a new node2Name function? If you don't want to do that, than you need to make sure clusterList.length === nodes.length at startup

I like the idea of using **name:fridendlyname**. At this time, I will add **clusterList.length === nodes.length** to the code. Once I make sure that the other changes are working fine then will incoporate **name:friendlyname** to the code.

> Also, I'm still unsure why you need to set _source.escluster as `egrep '(source|fields).escluster' *.js` returns no results other than in multies

This is being used in the Vueapp (`SessionDetail.vue`). In the current implementation, when a user click the `(+)` on the Sessions page then the UI fetches the details of that session using **session id** and **node**. However, in the new implementation, the UI will fetch session details using three parameters: **session id, node, and escluster.** As such, the MultiES will know which cluster to use to get the session details, instead of sending the same request to all the ES nodes. Please refer to the following UI codes for the details:

https://github.com/codesniffer/moloch/blob/0058a1cb824152d17b3fb088f7def2244a8e7b1b/viewer/vueapp/src/components/sessions/SessionDetail.vue#L625
https://github.com/codesniffer/moloch/blob/0058a1cb824152d17b3fb088f7def2244a8e7b1b/viewer/vueapp/src/components/sessions/SessionDetail.vue#L941


>
> minor stuff:
> 
> * I think in toplevel package.json, viewer:multiviewer should be renamed viewer:testmulties and instead of -n all it should use -n testmulties
> * the reason `isLocalViewRegExp=testmulties` is required is because we are abusing multiviewer, the design was never for multiviewer to have pcap locally, it was always supposed to reach out to real viewers, but I guess its ok for dev until something breaks. As long as the test suite doesn't use it.

Thanks for the suggestion. It does make sense.

> * how can i test the inactive part?

1. I am running two ES instances on my development machine one on port 9200 another on port 9201. Initially, all the clusters are up. 

2. Start MultiES and MultiViewer. 

3. Go to the the Multiviewer page and the Multiviewer shows the two clusters in the UI: Active (2) and Inactive (0)

4. Stop one of the ES instances.

5. The MultiES senses that an ES cluster is down (wait for a min or so)

6. Refresh the Multiviewer page. Now the ES menu will show: Active (1) and Inactive (1). If a user try to select the Inactive cluster it will be ignored.

7. Start the ES instance that was stopped in Step 4.

8. Repeat steps 5-6. Now, the UI will show Active (2) and Inactive (0).

Note that you can even stop both of the ES clusters (Step 4 or Step 7). In that case, the Multiviewer will display: Active (0) and Inactive (2)

**Review by 31453 — APPROVED** [review]
✨

**Comment by codesniffer** [comment]
@awick , I have completed making the major and minor changes. Please review and suggest when you have time.
